<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Simulating data with dependentsimr</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Simulating data with dependentsimr</h1>



<div id="setup" class="section level2">
<h2>Setup</h2>
<p>Install the required dependencies if not already installed:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># The simulator</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>  remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;tgbrooks/dependentsimr&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># DESeq2 - optional dependency only needed if doing RNA-seq data sets</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># Needed for type = &quot;DESeq2&quot;</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;DESeq2&quot;</span>)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># Two more dependencies that are used in some simulation modes</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co"># Needed for method = &quot;corpcor&quot;</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;corpcor&quot;</span>)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co"># Needed for method = &quot;spiked Wishart&quot;</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;sparsesvd&quot;</span>)</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co"># Extras needed for this vignette</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>)</span></code></pre></div>
<p>Load the libraries:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(dependentsimr)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">suppressWarnings</span>(<span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse, DESeq2)))</span></code></pre></div>
<p>Set a seed for reproducibility.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span></code></pre></div>
</div>
<div id="load-data" class="section level2">
<h2>Load data</h2>
<p>Simulations with <code>dependentsimr</code> start with a real data
set that should be mimicked. This data can be almost any shape,
including continuous, or discrete. Here, we will simulate RNA-seq, which
has discrete counts that we handle with a special “DESeq2” model. Other
data types can use normal, Poisson, or empirical (which uses only the
values present in the data set) types, or even a combination of
these.</p>
<p>We will use the following data set, which is a trimmed-down RNA-seq
data of 1000 genes on 12 samples, originating from GSE151923.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">head</span>(read_counts)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 6 × 13</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt;   gene_id     Arpp19_10_3 Arpp19_4_1 Arpp19_5_1 Arpp19_5_3 Arpp19_5_4 Arpp19_6_1</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   &lt;chr&gt;             &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; 1 ENSMUSG000…        2377       2162       2169       2650       2217       2552</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; 2 ENSMUSG000…         908        799        730        941        761        780</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; 3 ENSMUSG000…         316        342        373        373        325        405</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; 4 ENSMUSG000…         584        650        466        687        551        606</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; 5 ENSMUSG000…        2739       2531       2594       2988       2614       3081</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; 6 ENSMUSG000…         807        753        746        894        806        819</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; # ℹ 6 more variables: Arpp19_7_2 &lt;dbl&gt;, Arpp19_9_1 &lt;dbl&gt;, Dach1_11_1 &lt;dbl&gt;,</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; #   Dach1_2_2 &lt;dbl&gt;, Dach1_3_1 &lt;dbl&gt;, Dach1_5_3 &lt;dbl&gt;</span></span></code></pre></div>
<p>We first just format it as a matrix for use with
<code>dependentsimr</code> instead of as a tibble.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>count_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(read_counts[,<span class="dv">2</span><span class="sc">:</span><span class="fu">ncol</span>(read_counts)])</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">rownames</span>(count_matrix) <span class="ot">&lt;-</span> read_counts<span class="sc">$</span>gene_id</span></code></pre></div>
</div>
<div id="get-dependence-structure" class="section level2">
<h2>Get Dependence Structure</h2>
<p>The first step is to compute a ‘random structure’ for the data, which
encodes how the variables (genes, in this case) in the data set are
dependent upon each other. This computes a covariance matrix which
captures the dependence between genes, however, that matrix is not the
same as the usual sample covariance matrix. It also computes the shape
(such as mean and variance) of each variable individually.</p>
<p>There are three implemented methods (“pca”, “spiked Wishart”, and
“corpcor”) with different characteristics. We demonstrate here each.</p>
<p>First, the “pca” method takes a <code>rank</code> and includes only
gene-gene dependence up to that many dimensions. Typically small values
(2 or 3, for example) would be used for <code>rank</code>, similar to
when performing PCA, unless a large number of input samples are
available. Note that we put our <code>read_counts</code> data into a
<code>list()</code>. This is done since
<code>get_random_structure</code> also supports multiple data sets of
different ‘modes’ (for example, proteomics and transcriptomics done on
the same samples). We only have one ‘mode’ that we give the arbitrary
name of <code>data</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>rs_pca <span class="ot">&lt;-</span> <span class="fu">get_random_structure</span>(<span class="fu">list</span>(<span class="at">data=</span>count_matrix), <span class="at">method=</span><span class="st">&quot;pca&quot;</span>, <span class="at">rank=</span><span class="dv">2</span>, <span class="at">types=</span><span class="st">&quot;DESeq2&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="co">#&gt; converting counts to integer mode</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; estimating size factors</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co">#&gt; estimating dispersions</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; mean-dispersion relationship</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; final dispersion estimates</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<p>Second, the spiked Wishart method also takes a <code>rank</code>
value but uses it differently. Typically, this will be larger, up to the
total number of samples minus one, which is what we use here. For very
large sample counts, <code>rank</code> might need to be capped well
below the number of samples for computational efficiency reasons.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>rs_wishart <span class="ot">&lt;-</span> <span class="fu">get_random_structure</span>(<span class="fu">list</span>(<span class="at">data=</span>count_matrix), <span class="at">rank=</span><span class="dv">11</span>, <span class="at">method=</span><span class="st">&quot;spiked Wishart&quot;</span>, <span class="at">types=</span><span class="st">&quot;DESeq2&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">#&gt; converting counts to integer mode</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; estimating size factors</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; estimating dispersions</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; mean-dispersion relationship</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt; final dispersion estimates</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<p>Third, a method based off the <code>corpcor</code> package is
implemented. It takes no parameters but often underestimates the amount
of gene-gene dependence in the data.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>rs_corpcor <span class="ot">&lt;-</span> <span class="fu">get_random_structure</span>(<span class="fu">list</span>(<span class="at">data=</span>count_matrix), <span class="at">method=</span><span class="st">&quot;corpcor&quot;</span>, <span class="at">types=</span><span class="st">&quot;DESeq2&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; converting counts to integer mode</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#&gt; estimating size factors</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt; estimating dispersions</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; mean-dispersion relationship</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; final dispersion estimates</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; fitting model and testing</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; Estimating optimal shrinkage intensity lambda (correlation matrix): 0.7601 </span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; Estimating optimal shrinkage intensity lambda.var (variance vector): 1</span></span></code></pre></div>
<p>Lastly, we could generate data without gene-gene dependence for
comparison by removing the dependence from any of our random
structures.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>rs_indep <span class="ot">&lt;-</span> <span class="fu">remove_dependence</span>(rs_pca)</span></code></pre></div>
</div>
<div id="generate-data" class="section level2">
<h2>Generate Data</h2>
<p>Real RNA-seq data sets typically have variation between samples in
the total number of reads per sample. Here, we just copy the total
number of reads from the real data set to get realistic values. A
<code>library_size</code> of 1 corresponds to the average read depth of
the data used to generate the random structure. This step is only needed
when generating RNA-seq data with <code>type = &quot;DESeq2&quot;</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>actual_library_sizes <span class="ot">&lt;-</span> count_matrix <span class="sc">|&gt;</span> <span class="fu">apply</span>(<span class="dv">2</span>, sum)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>N_SAMPLES <span class="ot">&lt;-</span> <span class="dv">6</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>library_sizes <span class="ot">&lt;-</span> <span class="fu">sample</span>(actual_library_sizes <span class="sc">/</span> <span class="fu">mean</span>(actual_library_sizes), <span class="at">size=</span>N_SAMPLES, <span class="at">replace=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<div id="simulate-control-data" class="section level3">
<h3>Simulate Control Data</h3>
<p>Now we generate some ‘control’ samples, which have the same
expression levels (as well as gene-gene dependence) as was estimated
from the original data set. We’ll use <code>rs_pca</code> that we
generated above for example, but any of the methods could be used. Each
column is a separate simulated sample. Samples are independent of each
other in this model, however genes within a sample are dependent upon
each other.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>control_sim_data <span class="ot">&lt;-</span> <span class="fu">draw_from_multivariate_corr</span>(rs_pca, <span class="at">n_samples=</span>N_SAMPLES, <span class="at">size_factors=</span>library_sizes)<span class="sc">$</span>data</span></code></pre></div>
<p>Note that we used <code>$data</code> here because that was the name
we provided to <code>get_random_structure</code> and it was the only
mode of data that we had. If multiple modes were used to generate the
random structure, then we <code>draw_from_multivariate_corr</code>
returns a list with each of them.</p>
</div>
<div id="simulate-differentially-expressed-data" class="section level3">
<h3>Simulate Differentially Expressed Data</h3>
<p>To generate a set of ‘treatment’ samples which differ from the
‘control’ samples, we will have to modify the random structure to
reflect the desired expression values of the treatment group. We do this
by just choosing some genes at random and picking a fold-change for each
at random.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># fraction of all genes to modify in treatment</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>FRAC_DE <span class="ot">&lt;-</span> <span class="fl">0.1</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># Min and max value for the log fold change (LFC) for each DE gene chosen</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co"># non-DE genes will have a LFC of 0</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>LFC_MIN <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>LFC_MAX <span class="ot">&lt;-</span> <span class="fl">4.0</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>N_DE <span class="ot">&lt;-</span> <span class="fu">floor</span>(FRAC_DE <span class="sc">*</span> <span class="fu">nrow</span>(control_sim_data))</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co"># Pick which genes are DE</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>de_genes <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">nrow</span>(control_sim_data), N_DE)</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co"># Pick the LFCs of each of those genes</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>de_lfc <span class="ot">&lt;-</span> <span class="fu">runif</span>(N_DE, LFC_MIN, LFC_MAX) <span class="sc">*</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>), <span class="at">size=</span>N_DE, <span class="at">replace=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>Now, we create a new random structure object and modify it to have
the desired expression values. Information about the marginal
distribution of each gene (e.g., mean and variance but not dependence on
other genes) is stored in <code>rs$marginals</code> for a random
structure <code>rs</code>. The specifics depend upon the
<code>type</code> used in <code>get_random_structure</code>, but for
<code>DESeq2</code>, the <code>q</code> value determines the mean
expression, so we change that here.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>rs_treatment <span class="ot">&lt;-</span> rs_pca</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>rs_treatment<span class="sc">$</span>marginals<span class="sc">$</span>data<span class="sc">$</span>q[de_genes] <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>(de_lfc) <span class="sc">*</span> rs_treatment<span class="sc">$</span>marginals<span class="sc">$</span>data<span class="sc">$</span>q[de_genes]</span></code></pre></div>
<p>Lastly, we generate the data for the ‘treatment’ samples.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>treatment_sim_data <span class="ot">&lt;-</span> <span class="fu">draw_from_multivariate_corr</span>(rs_treatment, <span class="at">n_samples=</span>N_SAMPLES, <span class="at">size_factors=</span>library_sizes)<span class="sc">$</span>data</span></code></pre></div>
</div>
</div>
<div id="plot-the-generated-data" class="section level2">
<h2>Plot the generated data</h2>
<p>To demonstrate the data, we show a PCA plot of the control and
treatment groups.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>sim_data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(control_sim_data, treatment_sim_data)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(sim_data), <span class="at">center=</span><span class="cn">TRUE</span>, <span class="at">scale.=</span><span class="cn">TRUE</span>, <span class="at">rank.=</span><span class="dv">2</span>)</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>pca_points <span class="ot">&lt;-</span> <span class="fu">predict</span>(pca, <span class="fu">t</span>(sim_data))</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>pca_data <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">x=</span>pca_points[,<span class="dv">1</span>], <span class="at">y=</span>pca_points[,<span class="dv">2</span>]) <span class="sc">|&gt;</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;control&quot;</span>, N_SAMPLES), <span class="fu">rep</span>(<span class="st">&quot;treatment&quot;</span>, N_SAMPLES)))</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="fu">ggplot</span>(pca_data, <span class="fu">aes</span>(x, y, <span class="at">color=</span>group)) <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;PC1&quot;</span>, <span class="at">y =</span> <span class="st">&quot;PC2&quot;</span>)</span></code></pre></div>
<p><img role="img" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAw1BMVEUAAAAAAEgAAHQASHQASJwAdL8Av8QzMzNIAABIAEhIAHRISHRISJxInN9NTU1NTXJNTZJNcrBNksxyTU1yTXJyTZJycrBysOV0AAB0AEh0AHR0SAB0v790v/+STU2STXKSTZKSzP+cSACcSEicdACcnEicnHScv5yc3/+wck2wzLCw5f+/dAC/35y//9+////Mkk3M///fnEjf///lsHLl///r6+v4dm3/v3T/zJL/35z/5bD//7///8z//9///+X///9I0F0jAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAHuElEQVR4nO2cjXraNhhGP7KSpayh7bqtdB3tQraOtrQk9CcECPX9X9UsWcaWDX7tYEcCv+dpAP/I/XQiS4pkJAEpRFwH4DviOgDfEdcB+I64DsB3xHUAviOuA/AdqZ7kexp7K79d3wm1Z70cUj0JBQEoCEBBAAoCUBCAggAUBKAgQAP5v7m5oaCCE25uUoYoiIIoqG5BrIPqOKH2rJdDqiehIAAFASgIQEEACgJQEEkh1ZOwBAEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBHhgQfEz1LVnvRxSPUk66uYFbZ7Crz3r5ZDqSdJRU1Cw/P06CO7e9J9/i/dQUJrb/rPr4Mf7i+Drr/GuhxXkeR109fRDWILu/r6OSpImHTVbMS1m+ce34O7tx3Dr55Dmg/IJQScoQbfPY0GKZgrIkZQgBQVZLLfWQTuyk/5iXJsE/Xj/0m7FdmTH+mplmwTl+kEUVAQFAQqy08Y6KE8z+acgCqIgwwMIsuoyCspt263h0QqKMklBu7JrcklBDQhqRx20j6BtO2rPejmkepKy2bt/HdQSQTWfUHvWyyHVk1AQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCCAX4JWA+n815suev/KyXQ9EukGweLxWL8sHr8TefRlHz8HL2g1OA9/TqaLs1DMetTVPxtBZ8pZt9WC5ifTIJgpQUOzEb4kgoamOLVX0EzdQYte5GSuN9StNU7eV4MhBVHQTkHWXTXvjLfcYr3pAwvyik0lrZzElbTauR51WEkrVDP/zggKTDMfLM5E/noVNfP7+Tl8QYr5jr7Ong2YRqon8UqQqnZ23kUUFDKT3XcRBT0AUj0JBQEoCNBaQRMR1S2NOqcFtFXQJOxMzGV4YIKyJ38vOnkvQdHfdKtBl4KKBKlOFwVtFaRvsUD/bUNBKST5OOnofud6REEppHoSCgK0VpCuptcjOE3SVkGLs3P9PgFVUGsFTbrZDzs4cEH2GDUcsd4ISgb/D7QfZD+dT0HZYDLf79jkSsLOS9S3Wzz5R2SoPn/65fXJ1OwsLWg9OjefZqCaPiBB69FQDVhPztWrqmTVWy+ahp3EWyUFqfnbjKnDFxRlf/XnNFi9GqsNZUf/S+8sKcj0pFcD1M77KWhrHbR48sVoWl9mBKV2lhWkZ5OkA8e5PRVkpzE5qrUEleWABKk6KFSQVDeJoOp10DEKsluxSFD4x/gnZaVqK5YMuR6ToP2RzSc1HjQrY6ilgtTtGr0gWirI/C0P+kC5mCgIxERBICYK0pRcgerIBUlMrinLrcTZRkFFlF4Fr62CPF6J0w9BHq/E6Ycgj1fi9EMQ6yBA2ZU4WyvoYPtBWwczikY48sfKCdodkz+CTk9PKaggmNPTtKHoYGaKR48p63mfz5fvRM7n4U+QmxPKGDpqQUFqimfWVUPQZsxVfy3xrKsLTHQsmRPK/AfHL6inB+f1IH2gB/CVoMuxmtJQB8yxZMD6WAXtqoO0oIGZrtGjyhlB0bEWCLLT2IJM6RkM41ssJSg61m5BqSke9T1ES1CQnRNqlaDUFI+6w2YiP70ehjs/pwRFx5I5oWNtxbYKqgGpnoSCABQEoCAABQG8EtQ4Uj0JBQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASiIpJDqSViCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCAFG85kvGFJQn9qMNUVAeCgJQEOB7bKjG/KeWTdj8Ag5cUMXff/EJ6YU3YmH2DgqioGMVpBd3a3yJrsOtg277z659Xqq0caT48NXTD2EJ8niZwMYRdIIS4/FSpY0j6AQlyOOlShtHdh656vdVtWOXIAUFWSzvXwfplqgVgu63VGnUl2mFoPv1g1oiKA8FAUpnrx11UJ57ZY+CKGgDBQEoCEBBAAoCUBCgMDvWqs71CIrnT9wg1ZMUZcdeF7wWQZsZODdI9SQUBKAgAMpOtfy3rA5iKxZQEISCABQEoCAABQEoCEBB+wBn7vc/4WGRmq9HQQAKAlBQ2xDXAfiOuA7Ad8R1AL4jrgPwHanxWssX/f5F5mEim4JDpdI7QOq7lHqKcfnbR/uhaouCQ6XSu0Dqu9StytXVhf1Ao0XBoVLpXSD1Xi4sBfYjsRYFh0qld4HUejX1tKf9ULVFwaFS6V0g9Vwmeqj67s3LwmKCCgdK7wKp8VrLF2EbVFTRgOoFpneB1HepKH+Zh6otCg6VSu8Cqe9SX/uKi3v3g0qkd4C4DsB3xHUAviOuA/AdcR2A74jrAHxHXAfgO9Ls5dcjUXTGamO2+RSsXo2b/Y9rQ5q9/Hp0rt5mJ9Pw46MvStJQ7Zh0KEhjBK0Gw2Ci/ISGwrfVQCgoIhGkHMXMu4vHFKQxgiYnU1sJBRlMJR1WQYveNLWfggymBAVZJRRkSATFdVDUwFOQIRGkm68gavEpaENKkNUPoiBDSlDYliU9aQo6GsR1AL4jrgPwHXEdgO+I6wB8R1wH4DviOgDfEdcB+I64DsB3xHUAvvM/B10cxK9Pso0AAAAASUVORK5CYII=" /><!-- --></p>
<p>As expected, the control and treatment group separate nicely, but
there is also a substantial component of variation (PC1) which both
groups have. This arises from the dependence between genes, which is the
same in both groups in our simulation.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
